### DISCUSSION TITLE
Downloading files is super slow

### DISCUSSION THREAD

**Discussion by ChronoDK** [fastapi/fastapi#4377] [Questions] [answered]
Hi,

I'm hosting a webgame of about 30mb. The files are located in a mounted directory:

`app.mount("/public", StaticFiles(directory="public", html=True), name="public")`

And the container html is served like this:

`@app.get("/", response_class=HTMLResponse)`

App is started like this:

`CMD gunicorn app.main:app -w 4 -k uvicorn.workers.UvicornWorker`

Just simple FastAPI stuff. The problem is, sometimes we will update the game and it will ask players to reload. 1000+ players do this at the same time, causing the download to be super slow. Looking at the server it does not seem to be capped by bandwidth, but the CPU is maxed out by gunicorn workers.

Is there a way to solve this problem? Why are the gunicorn workers so busy?

**Comment by jgould22** [comment, marked as answer]
The static files in FastAPI is just a reexport of the feature from starlette.

You can see how the file response works in the source here https://github.com/encode/starlette/blob/ad495c7ac538e8583fb6aac0aa2111dc8d8458eb/starlette/staticfiles.py and here https://github.com/encode/starlette/blob/ad495c7ac538e8583fb6aac0aa2111dc8d8458eb/starlette/responses.py

So for each of your requests it is opening the file and reading it back which is going to back things up if you have 1000 requests at once.

Do all the clients need to update at once? If not maybe have them do a random offset to ease the load on the server. 

Another option that might help is to make your own endpoint that caches the file and serves it which decrease the load on the disk.

A third option is to place something in front of your python application that caches requests, like some proxy, or some service like CloudFlare.

**Reply by ChronoDK** [reply]
Not forcing the update does indeed work, but sometimes the update is urgent. The server is behind a Cloudflare tunnel already, but I don't think the caching works instantly.

**Reply by jgould22** [reply]
Be mindful that they only cache some file extension types and there are limits to file sizes based on plans 

https://developers.cloudflare.com/cache/about/default-cache-behavior#default-cached-file-extensions

**Reply by ChronoDK** [reply]
Whoa, thanks a lot - I added Cloudflare page rules to cache the correct extensions, and it lowered my bandwidth usage hugely, but also CPU usage is a lot lower now. Still needs more testing, but this could be the solution. 

Maybe I should actually read more about all the things Cloudflare can do ;-)
